lib/active_model/serializer/adapter/json_api/error.rb

```diff
@@ -0,0 +1,53 @@
+class ActiveModel::Serializer::Adapter::JsonApi::Error < ActiveModel::Serializer::Adapter
+  def serializable_hash(options = nil)
+    options ||= {}
+    @result = []
+    # TECHDEBT: clean up single vs. collection of resources
+    if serializer.object.respond_to?(:each)
+      @result = collection_errors.flat_map { |collection_error|
+        collection_error.flat_map { |attribute_name, attribute_errors|
+          attribute_error_objects(attribute_name, attribute_errors)
+        }
+      }
+    else
+      @result = object_errors.flat_map { |attribute_name, attribute_errors|
+        attribute_error_objects(attribute_name, attribute_errors)
+      }
+    end
+    { root => @result }
+  end
+
+  def fragment_cache(cached_hash, non_cached_hash)
+    JsonApi::FragmentCache.new().fragment_cache(root, cached_hash, non_cached_hash)
+  end
+
+  def root
+    "errors".freeze
+  end
+
+  private
+
+  # @return [Array<symbol, Array<String>] i.e. attribute_name, [attribute_errors]
+  def object_errors
+    cache_check(serializer) do
+      serializer.object.errors.messages
+    end
+  end
+
+  def collection_errors
+    cache_check(serializer) do
+      serializer.object.flat_map { |elem|
+        elem.errors.messages
+      }
+    end
+  end
```

Comment:
I wonder if this can be made compatible with https://github.com/rails-api/active_model_serializers/pull/1114/files?diff=split#r38722592